dds/DCPS/Xcdr2ValueWriter.cpp

```diff
@@ -0,0 +1,836 @@
+      size_cache_.push_back(uint64_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << value;
+  }
+}
+
+bool Xcdr2ValueWriter::write_float32(ACE_CDR::Float value)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    primitive_serialized_size(encoding_, size_states_.top().total_size, value);
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      size_cache_.push_back(float32_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << value;
+  }
+}
+
+bool Xcdr2ValueWriter::write_float64(ACE_CDR::Double value)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    primitive_serialized_size(encoding_, size_states_.top().total_size, value);
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      size_cache_.push_back(float64_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << value;
+  }
+}
+
+bool Xcdr2ValueWriter::write_float128(ACE_CDR::LongDouble value)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    primitive_serialized_size(encoding_, size_states_.top().total_size, value);
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      size_cache_.push_back(float128_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << value;
+  }
+}
+
+bool Xcdr2ValueWriter::write_fixed(const ACE_CDR::Fixed& /*value*/)
+{
+  return true;
+}
+
+bool Xcdr2ValueWriter::write_char8(ACE_CDR::Char value)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    primitive_serialized_size_char(encoding_, size_states_.top().total_size);
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      size_cache_.push_back(char8_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << ACE_OutputCDR::from_char(value);
+  }
+}
+
+bool Xcdr2ValueWriter::write_char16(ACE_CDR::WChar value)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    primitive_serialized_size_wchar(encoding_, size_states_.top().total_size);
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      size_cache_.push_back(char16_cdr_size);
+    }
+    return true;
+  } else {
+    return *ser_ << ACE_OutputCDR::from_wchar(value);
+  }
+}
+
+bool Xcdr2ValueWriter::write_string(const ACE_CDR::Char* value, size_t length)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    size_t& size = size_states_.top().total_size;
+    primitive_serialized_size_ulong(encoding_, size);
+    if (value) {
+      size += length + 1; // Include null termination
+    }
+
+    if (size_states_.top().extensibility == MUTABLE) {
+      // It's safe to do this since before every member of a mutable type,
+      // the total_size variable is set to zero.
+      size_cache_.push_back(size);
+    }
+    return true;
+  } else {
+    return *ser_ << value;
+  }
+}
+
+bool Xcdr2ValueWriter::write_wstring(const ACE_CDR::WChar* value, size_t length)
+{
+  if (mode_ == SERIALIZATION_SIZE_MODE) {
+    size_t& size = size_states_.top().total_size;
+    primitive_serialized_size_ulong(encoding_, size);
+    if (value) {
```

Comment:
If value is null isn't this an error?  Not sure if it's one that needs to be detected here.

Reply:
Yes, I think it's an error. `DynamicDataImpl` hasn't checked for a null string in `set_string_value` and `set_wstring_value`. I think it should do that.